clang/test/CodeGen/switch-case-overflow.c

```diff
@@ -0,0 +1,37 @@
+// RUN: split-file %s %t
+// RUN: python %t/gen.py %t/switch-overflow.c %t/tmp.c && %clang_cc1 -emit-llvm %t/tmp.c -o - | FileCheck %t/tmp.c
+
+//--- gen.py
```

Comment:
I guess 32000 is small enough that a test won't be that expensive, but a python script seems like overkill.  Maybe use a bit of C preprocessor trickery like the following:

```
#define CASES1(n) case n: case n+1: case n+2 case n+3:
#define CASES2(n) CASES1(n) CASES1(n+4) CASES1(n+8) CASES1(n+12)
#define CASES3(n) CASES2(n) CASES2(n+16) CASES2(n+32) CASES2(n+48)
void foo() {
  switch (1337) {
  CASES3(0)
    break;
  }
}
```

Reply:
Yeah, we've had to add a bunch of recursion hacks around `CaseStmt` throughout the compiler over the years, so another one would hardly be unprecedented.

Frankly, we should probably just change the representation in the AST. There are a lot of things in the C grammar that are formally recursive that we fold into flat arrays, like call arguments. This is less self-evident than most of those, but these ubiquitous recursion hacks are a real problem, and we can easily avoid it with a better representation. I feel that argument is pretty compelling. We could just collect sequences of adjacent `case`s into a single `Stmt`. There might be places that use `CaseStmt*` today in a way that assumes it represents a single case, but we can just update those to store a `CaseStmt*` and an index.

Reply:
On my local machine, `32,000` seems to work alright and it seems to crash at somewhere around `60,000`. I increased the test size bc of this, but I can change it back if that's too much.

> Frankly, we should probably just change the representation in the AST

Should I do this as part of this PR?